fiftyone/core/dataset.py

```diff
@@ -2884,6 +2884,96 @@ def delete_group_slice(self, name):
 
         self.save()
 
+    def map_samples(
+        self,
+        map_func,
+        map_func_args=None,
+        aggregate_func=None,
+        num_workers=4,
+        progress=False,
+        autosave=False,
+        batch_size=None,
+        batching_strategy=None,
+        skip_failures=True,
+        warn_failures=False,
+    ):
+        """Maps a function over the samples in the dataset.
+
+        Args:
+            map_func: a function that accepts a :class:`fiftyone.core.sample.Sample` as
+                an input and returns a result
+            map_func_args (None): additional arguments to pass to the map_func
+            aggregate_func (None): an optional function that accepts a list of
+                the results of the mapping operation and returns the final
+                result. By default, the results are returned as a list
+            num_workers (4): the number of worker threads to use
+            progress (False): whether to render a progress bar (True/False)
+            autosave (False): whether to automatically save the results
```

Comment:
Note that there's no point in `map_samples(autosave=False)`. If you are applying changes to samples, you need to be saving them.

The only reason that `iter_samples(autosave=False)` makes sense is that the user is assumed to be calling `sample.save()` inside the for loop manually.

Reply:
Is there a scenario where `map_func` does not modify the samples? We can just compute some large scale statistics and aggregate the results at the end without modifying samples.

Reply:
Yeah I think that's right Minh. We should make the default `autosave=True`, but should leave it in, because if the caller says they don't need to save samples then we can elide that whole process.